src/cable/server.cr

```diff
@@ -0,0 +1,148 @@
+        @channels[identifier] = [] of Cable::Channel
+      end
+
+      @channels[identifier] << channel
+
+      request = Redis::Request.new
+      request << "subscribe"
+      request << identifier
+      redis_subscribe._connection.send(request)
+    end
+
+    def unsubscribe_channel(channel : Channel, identifier : String)
+      if @channels.has_key?(identifier)
+        @channels[identifier].delete(channel)
+
+        if @channels[identifier].size == 0
+          request = Redis::Request.new
+          request << "unsubscribe"
+          request << identifier
+          redis_subscribe._connection.send(request)
+
+          @channels.delete(identifier)
+        end
+
+      else
+        request = Redis::Request.new
+        request << "unsubscribe"
+        request << identifier
+        redis_subscribe._connection.send(request)
+      end
+    end
+
+    def publish(channel, message)
+      redis_publish.publish("#{channel}", message)
+    end
+
+    def send_to_channels(channel, message)
+      parsed_message = JSON.parse(message)
+
+      @channels[channel].each do |channel|
+        Cable::Logger.info "#{channel.class} transmitting #{parsed_message} (via streamed from #{channel.stream_identifier})"
+        channel.connection.socket.send({
+          identifier: channel.identifier,
+          message:    parsed_message,
+        }.to_json)
+      end
+    rescue IO::Error
+    end
+
+    def debug
+      Cable::Logger.debug "-" * 80
+      Cable::Logger.debug "Some Good Information"
+      Cable::Logger.debug "Connections"
+      @connections.each do |k, v|
+        Cable::Logger.debug "Connection Key: #{k}"
+      end
+      Cable::Logger.debug "Channels"
+      @channels.each do |k, v|
+        Cable::Logger.debug "Channel Key: #{k}"
+        Cable::Logger.debug "Channels"
+        v.each do |channel|
+          Cable::Logger.debug "From Channel: #{channel.connection.connection_identifier}"
+          Cable::Logger.debug "Params: #{channel.params}"
+          Cable::Logger.debug "ID: #{channel.identifier}"
+          Cable::Logger.debug "Stream ID:: #{channel.stream_identifier}"
+        end
+      end
+      Cable::Logger.debug "-" * 80
+    end
+
+    def shutdown
+      request = Redis::Request.new
+      request << "unsubscribe"
+      redis_subscribe._connection.send(request)
+      redis_subscribe.close
+      redis_publish.close
+      connections.each do |k, v|
+        v.close
+      end
+    end
+
+    private def process_subscribed_messages
+      server = self
+      spawn do
+        while received = fiber_channel.receive
+          channel = received[0]
+          message = received[1]
+          server.send_to_channels(channel, message)
+        end
+      end
+    end
+
+    private def subscribe
+      spawn do
+        redis_subscribe.subscribe("_internal") do |on|
+          on.message do |channel, message|
+            if channel == "_internal" && message == "debug"
+              puts self.debug
```

Comment:
Since this is in a spawn, would you ever actually see this?

Reply:
yes! :D

I was opening a  `redis-cli` and `publish _internal "debug"` to get all the information about current state of the server... was pretty neat usage of "_internal" channel 😉

Reply:
ah interesting 😄